ldk-sample/src/main/java/com/getlipa/ldk/sample/LDKSample.java

```diff
@@ -0,0 +1,360 @@
+        final var channelManagerFile = new File("data/channel-manager.dat");
+        byte[] serializedChannelManager = null;
+        if (channelManagerFile.exists() && channelManagerFile.isFile()) {
+            serializedChannelManager = Files.readAllBytes(channelManagerFile.toPath());
+        }
+
+        ChannelManagerConstructor channelManagerConstructor;
+        if (serializedChannelManager == null) {
+            final var currentBlockHeight = chainBackend.blockHeight(); //126;
+            final var latestBlockHash = chainBackend.blockHash(currentBlockHeight);
+            channelManagerConstructor = new ChannelManagerConstructor(
+                    Network.LDKNetwork_Regtest,
+                    UserConfig.with_default(),
+                    latestBlockHash,
+                    currentBlockHeight,
+                    keyManager.as_KeysInterface(),
+                    feeEstimator,
+                    chainMonitor,
+                    null,
+                    txBroadcaster,
+                    logger
+            );
+        } else {
+            channelManagerConstructor = new ChannelManagerConstructor(
+                    serializedChannelManager,
+                    channelMonitors,
+                    keyManager.as_KeysInterface(),
+                    feeEstimator,
+                    chainMonitor,
+                    filter,
+                    null,
+                    txBroadcaster,
+                    logger
+            );
+        }
+        final ChannelManager channelManager = channelManagerConstructor.channel_manager;
+
+        // Step 6
+        final var channelManagerPersister = new ChannelManagerConstructor.ChannelManagerPersister() {
+
+            @Override
+            public void handle_event(final Event e) {
+                final var params = NetworkParameters.fromID(ID_REGTEST);
+                if (e instanceof Event.FundingGenerationReady) {
+                    var event = (Event.FundingGenerationReady) e;
+                    final var transaction = new Transaction(params);
+                    final var input = new TransactionInput(params, transaction, new byte[0]);
+                    input.setWitness(new TransactionWitness(2));
+                    input.getWitness().setPush(0, new byte[]{0x1});
+                    transaction.addInput(input);
+                    final var script = new Script(event.output_script);
+                    final var value = Coin.SATOSHI.multiply(event.channel_value_satoshis);
+                    transaction.addOutput(value, script);
+                    channelManager.funding_transaction_generated(event.temporary_channel_id, transaction.bitcoinSerialize());
+                } else if (e instanceof Event.PaymentReceived) {
+                    var event = (Event.PaymentReceived) e;
+                    System.out.printf("Payment of %s SAT received.%n", event.amt);
+                    channelManager.claim_funds(event.payment_preimage);
```

Comment:
Hmm, which BOLT compliance questions do you have here? I believe ~all BOLT compliance (and several other issues like privacy concerns etc) are handled internal to LDK, the only things you need to be concerned about are those explicitly stated in the documentation for `PaymentReceived` and/or `claim_funds`.

Reply:
This was actually the only point I noticed that BOLT compliance may be broken by the LDK user. However, it made me wondering whether there are other parts I have to make sure my implementation follows the spec. If this is really the only place LDK users have to ensure BOLT compliance by themselves, I would appreciate having a default implementation for this event which is BOLT compliant.

At this point I am also wondering, why LDK users have to provide own event handler implementations at all. As mentioned, I really like the possibility to add custom logic when specific events occur. But isn't it that the basic stuff will always be the same? I think some sort of default implementations relying on the already existing interfaces (e.g. for transaction broadcast) and maybe an additional one for payment persistence (similar to the `Persist` interfaces for channels) would help a lot.

Reply:
> I hope there isn't a way you can fall out of BOLT compliance with LDK unless you try very, very hard to :).

That's good to hear - thanks for making this clear!

> Just naively claiming the payment and doing nothing else may be ok for a sample, but any real application will need to do more to hook it up to their UX.

This is actually more or less of my point: Naively claiming the payment may be ok for a sample but for a real app there are at least some checks required before the claiming. Some of them are even defined by a BOLT, which is why I would appreciate having an implementation provided by LDK that performs those checks (e.g. received amount should not be greater than twice of the requested amount). That's why I wanted to suggest default implementations for the event handlers (see also below).

> Payment persistence we could provide a default, but I'd be a little surprised if any real-world application wouldn't want to store additional metadata attached to nearly each payment - invoice description, etc, and then hook it up explicitly to their UX, where it maybe belongs better.

I guess you got me wrong at this point. I was not talking about having a default implementation for payment persistence but about default implementations for the event handlers. For the `PaymentReceived` event this would require having an interface for payment persistence. This could look as follows: 
```
interface PaymentPersistence {
  PaymentRequest findPaymentRequest(byte[] paymentHash);
}

interface PaymentRequest {
  byte[] getPreimage();
  int getAmountSat();
}
```
With such an interface, LDK could handle the mentioned checks in the case of an incoming payment event. It could verify whether the requested amount is sufficient but not more than twice as the requested amount. The way payments are actually stored would have to be defined by the LDK user by implementing the above interfaces. 

For the other event handlers, similar default implementations would also be useful. I think they could even be realized by relying on the already existing interfaces for e.g. transaction broadcast. However, this does not necessarily mean that a default implementation of e.g. transaction broadcast has to be provided by LDK.

Reply:
> This is actually more or less of my point: Naively claiming the payment may be ok for a sample but for a real app there are at least some checks required before the claiming. Some of them are even defined by a BOLT, which is why I would appreciate having an implementation provided by LDK that performs those checks (e.g. received amount should not be greater than twice of the requested amount). That's why I wanted to suggest default implementations for the event handlers (see also below).

The BOLT-prescribed checks (and privacy-important checks) are already implemented in LDK before generating the `PaymentReceived` event. eg the receive amount check and payment secret checks are done based on the inbound payment calls (eg https://docs.rs/lightning/0.0.100/lightning/ln/channelmanager/struct.ChannelManager.html#method.create_inbound_payment).

Reply:
In general, I'm not sure that users will need to perform their own *checks* before claiming HTLCs, but I do figure they'll need to persist additional information to track payments. eg you probably want a notification, you probably want to match a payment to an invoice generated in the past and update some local database, etc.

Reply:
> The BOLT-prescribed checks (and privacy-important checks) are already implemented in LDK before generating the PaymentReceived event. 

Ah ok, that was not clear to me - then I see no need for a "default" implementation. So this means whenever a `PaymentReceived` event occurs, the payment meets all BOLT requirements? I guess this also means that LDK declines a payment if for example the sent amount is not sufficient?

However, this raises more questions: If LDK performs BOLT-prescribed checks itself, this means that LDK needs to know which payment requests were generated when a payment is incoming. How is this persisted? Is it part of the channel manager or is it just in-memory?

It was the following part of the rust docs which made me believe the LDK user has to make sure themselves that the paid amount is correct:
> Note that if the preimage is not known or the amount paid is incorrect, you should call ChannelManager::fail_htlc_backwards to free up resources for this HTLC and avoid network congestion. The amount paid should be considered ‘incorrect’ when it is less than or more than twice the amount expected.-

-- https://docs.rs/lightning/0.0.100/lightning/util/events/enum.Event.html#variant.PaymentReceived

So this part is actually wrong / outdated?

Reply:
> Ah ok, that was not clear to me - then I see no need for a "default" implementation. So this means whenever a PaymentReceived event occurs, the payment meets all BOLT requirements? I guess this also means that LDK declines a payment if for example the sent amount is not sufficient?

We do all the required checks now, yes, including amount, if the user provided LDK one (though the privacy concerns over amounts no longer apply with required payment secrets).

> However, this raises more questions: If LDK performs BOLT-prescribed checks itself, this means that LDK needs to know which payment requests were generated when a payment is incoming. How is this persisted? Is it part of the channel manager or is it just in-memory?

See https://docs.rs/lightning/0.0.100/lightning/ln/channelmanager/struct.ChannelManager.html#method.create_inbound_payment for where we accept that information. It is persisted as a part of the ChannelManager object.

> So this part is actually wrong / outdated?

Yes, thanks for reporting, fixed at https://github.com/rust-bitcoin/rust-lightning/pull/1079